pkg/rbd/rbd_util.go

```diff
@@ -112,36 +119,93 @@ type rbdSnapshot struct {
 
 var (
 	supportedFeatures = sets.NewString("layering")
+
+	// large interval and timeout, it should be longer than the maximum
+	// time an operation can take (until refcounting of the connections is
+	// available)
+	connPool = util.NewConnPool(15*time.Minute, 10*time.Minute)
 )
 
 // createImage creates a new ceph image with provision and volume options.
 func createImage(ctx context.Context, pOpts *rbdVolume, volSz int64, cr *util.Credentials) error {
-	var output []byte
-
-	image := pOpts.RbdImageName
 	volSzMiB := fmt.Sprintf("%dM", volSz)
+	options := librbd.NewRbdImageOptions()
 
-	logMsg := "rbd: create %s size %s format 2 (features: %s) using mon %s, pool %s "
+	var err error
+	imageFormat := rbdImageFormat2
+	if pOpts.ImageFormat != "" {
+		imageFormat, err = strconv.Atoi(pOpts.ImageFormat)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert ImageFormat (%v) to integer", pOpts.ImageFormat)
+		}
+	}
+
+	err = options.SetUint64(librbd.RbdImageOptionFormat, uint64(imageFormat))
+	if err != nil {
+		return errors.Wrapf(err, "failed to set ImageFormat to %v", imageFormat)
+	}
+
+	logMsg := "rbd: create %s size %s format %d (features: %s) using mon %s, pool %s "
 	if pOpts.DataPool != "" {
 		logMsg += fmt.Sprintf("data pool %s", pOpts.DataPool)
+		err = options.SetString(librbd.RbdImageOptionDataPool, pOpts.DataPool)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set data pool")
+		}
 	}
 	klog.V(4).Infof(util.Log(ctx, logMsg),
-		image, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
+		pOpts.RbdImageName, volSzMiB, imageFormat, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
 
-	args := []string{"create", image, "--size", volSzMiB, "--pool", pOpts.Pool, "--id", cr.ID, "-m", pOpts.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature", pOpts.ImageFeatures}
+	if pOpts.ImageFeatures != "" {
+		features := imageFeaturesToUint(ctx, pOpts.ImageFeatures)
+		err := options.SetUint64(librbd.RbdImageOptionFeatures, features)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set image features")
+		}
+	}
 
-	if pOpts.DataPool != "" {
-		args = append(args, "--data-pool", pOpts.DataPool)
+	ioctx, err := pOpts.getIoctx(cr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get IOContext")
 	}
-	output, err := execCommand("rbd", args)
+	defer ioctx.Destroy()
 
+	err = librbd.CreateImage(ioctx, pOpts.RbdImageName, uint64(volSz*util.MiB), options)
 	if err != nil {
-		return errors.Wrapf(err, "failed to create rbd image, command output: %s", string(output))
+		return errors.Wrapf(err, "failed to create rbd image")
 	}
 
 	return nil
 }
 
+func (rv *rbdVolume) getIoctx(cr *util.Credentials) (*rados.IOContext, error) {
```

Comment:
Functions that don't use its receiver proably should just be top-level helper functions without being glued to a particular type.

Reply:
I think this does use its receiver? The functions I modified got their 'receiver' as an argument. The partial object-oriented, partial procedural approach is rather ugly. Hopefully we can move to a more object-oriented approach with follow-up changes.